File: android/src/main/kotlin/com/cloudflare/realtimekit/flutter/view/plugin/RtkPluginViewFactory.kt
package com.cloudflare.realtimekit.flutter.view.plugin
import android.content.Context
import com.cloudflare.realtimekit.RealtimeKitClient
import com.cloudflare.realtimekit.flutter.RtkClientProvider
import com.cloudflare.realtimekit.plugins.RtkPlugin
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory

class RtkPluginViewFactory(private val mobileClient: RealtimeKitClient) :
    PlatformViewFactory(StandardMessageCodec.INSTANCE) {
    override fun create(context: Context?, viewId: Int, args: Any?): PlatformView {
        // Live client (rebuilt per meeting), not the captured one.
        val mobileClient = RtkClientProvider.requireRealtimeClient()
        val creationParams = args as Map<String?, Any?>
        val dytePlugin: RtkPlugin =
            mobileClient.plugins.all.find { it.id == creationParams["id"] }!!
        if (context == null) {
            throw AssertionError("context for DytePluginView is null!")
        }
        return RtkPluginPlatformView(dytePlugin, context)
    }

}

File: ios/realtimekit_core_ios/Sources/realtimekit_core_ios/view/RtkPluginViewFactory.swift
import Flutter
import RealtimeKitFlutterCoreKMM

class RtkPluginViewFactory: NSObject, FlutterPlatformViewFactory {
    // Read the live client dynamically so views created after a rejoin (which
    // rebuilds the native client) bind to the current client, not a stale one.
    var mobileClient: CoreRealtimeKitClient { RtkClientProvider.shared.core }

    init(client _: CoreRealtimeKitClient) {}

    func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
        return FlutterStandardMessageCodec.sharedInstance()
    }

    func create(withFrame frame: CGRect, viewIdentifier viewId: Int64, arguments args: Any?) -> FlutterPlatformView {
        guard let creationParams = args as? [String: Any?],
              let id = creationParams["id"] as? String,
              let dytePlugin = mobileClient.plugins.all.first(where: { $0.id == id })
        else {
            fatalError("Failed to find DytePlugin in mobileClient.plugins")
        }
        return DytePluginView(frame: frame, viewIdentifier: viewId, plugin: dytePlugin)
    }
}
